### PR TITLE
 Simplify s3 functions: delete_s3_bucket.

### PR DESCRIPTION
**Pull request by DailyDreaming**
Making a series of many smaller PRs to simplify some of the AWS jobstore logic (like having retries on the imported functions themselves, rather than retyping the retries ad nauseam for every usage) out into generic callable AWS functions.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 *  Simplify s3 functions: delete_s3_bucket.

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

File: src/toil/lib/aws/s3.py [hunk 1]
```diff
@@ -82,3 +85,60 @@ def create_s3_bucket(
...
+    quiet: bool = True
+) -> None:
+    """
+    Delete the bucket with 'bucket_name'.
+
+    Note: 'quiet' is False when used for a clean up utility script (contrib/admin/cleanup_aws_resources.py)
+         that prints progress rather than logging.  Logging should be used for all other internal Toil usage.
+    """
+    logger.debug("Deleting bucket '%s'.", bucket_name)
+    printq(f'\n * Deleted s3 bucket successfully: {bucket_name}\n\n', quiet)
```
Comment (adamnovak) [inline src/toil/lib/aws/s3.py:103]: This is a lie. The bucket is still there at this point.

File: src/toil/lib/aws/s3.py [hunk 2]
```diff
@@ -82,3 +85,60 @@ def create_s3_bucket(
...
+
+        paginator = s3_client.get_paginator('list_object_versions')
+        for response in paginator.paginate(Bucket=bucket_name):
+            # Versions and delete markers can both go in here to be deleted.
+            # They both have Key and VersionId, but there's no shared base type
+            # defined for them in the stubs to express that. See
+            # <https://github.com/vemel/mypy_boto3_builder/issues/123>. So we
+            # have to do gymnastics to get them into the same list.
+            to_delete: List[Dict[str, Any]] = cast(List[Dict[str, Any]], response.get('Versions', [])) + \
+                                              cast(List[Dict[str, Any]], response.get('DeleteMarkers', []))
```
Comment (adamnovak) [inline src/toil/lib/aws/s3.py:123]: Maybe you can fix my code to use `Union` here?

**Review by adamnovak — CHANGES_REQUESTED** [review]
This looks almost great, except I think the non-logging progress report will tell lies.

**Comment by adamnovak** [comment]
Are we going to merge this into the other PR's branch and *then* merge that?

**Comment by DailyDreaming** [comment]
@adamnovak That's one way of merging it, but merging them into master in the proper order also works (and which I'll do).  Comparing against the other branch just makes the PR reviews separate and simpler.  I like this method for PR creation when the parts are distinct.

File: src/toil/lib/aws/s3.py (hunk 1 above)
Comment (adamnovak) [reply src/toil/lib/aws/s3.py:103]:
```suggestion
    printq(f'\n * Deleting s3 bucket: {bucket_name}\n\n', quiet)
```

**Comment by adamnovak** [comment]
I'm going to make the tiny change I think this needs and see if I can get it and the other one merged.

**Review by adamnovak — APPROVED** [review]

**Pull request merged**